### PR TITLE
feat: overwrite mount pod resources with PVC annotations

### PR DESCRIPTION
**Pull request by zxh326**

### CODE REVIEW

**Comment by timfeirg** [comment]
since this pr introduces annotation based resource declaration, there's no reason to continue using [`volumeAttributes`](https://juicefs.com/docs/csi/guide/resource-optimization/#mount-pod-resources), correct?

if that's the case, we should recommend annotation as the first option for resources declaration

**Comment by zxh326** [comment]
> since this pr introduces annotation based resource declaration, there's no reason to continue using [`volumeAttributes`](https://juicefs.com/docs/csi/guide/resource-optimization/#mount-pod-resources), correct?
> 
> if that's the case, we should recommend annotation as the first option for resources declaration

Now we have two ways to customize the mount pod resource.

1. Static PV, volumeAttributes

2. Using parameters of storageClass

these two methods can be modified later.

We can consider not abandoning these two ways for now, and suggest customizing the mount pod resource using pvc annotations in the documentation.

**Comment by timfeirg** [comment]
the old ways will certainly be preserved, I'll remember to create a docs pr after this is released

File: pkg/juicefs/juicefs.go [hunk 1]
```diff
@@ -339,6 +343,38 @@ func (j *juicefs) JfsMount(ctx context.Context, volumeID string, target string,
...
+		klog.Errorf("Get pvc %s/%s error: %v", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name, err)
+		return err
+	}
+
+	cpuLimit := pvc.Annotations[config.MountPodCpuLimitKey]
+	memoryLimit := pvc.Annotations[config.MountPodMemLimitKey]
+	cpuRequest := pvc.Annotations[config.MountPodCpuRequestKey]
+	memoryRequest := pvc.Annotations[config.MountPodMemRequestKey]
+
+	if cpuLimit != "" {
```
Comment (davies) [inline pkg/juicefs/juicefs.go]: We can use a for-loop here

**Review by davies — APPROVED** [review]

File: pkg/controller/pod_driver.go [hunk 2]
```diff
@@ -375,6 +375,9 @@ func (p *PodDriver) podDeletedHandler(ctx context.Context, pod *corev1.Pod) erro
 				}
 				controllerutil.AddFinalizer(newPod, config.Finalizer)
 				klog.Infof("Need to create pod %s %s", pod.Name, pod.Namespace)
+				if err := p.OverwirteMountPodResourcesWithPVC(ctx, newPod); err != nil {
+					klog.Errorf("overwrite mount pod resources wiht pv error %v", err)
```
Comment (timfeirg) [inline pkg/controller/pod_driver.go]: typo

File: pkg/juicefs/juicefs.go [hunk 3]
```diff
@@ -318,6 +318,10 @@ func (j *juicefs) JfsMount(ctx context.Context, volumeID string, target string,
 	if err := j.validTarget(target); err != nil {
 		return nil, err
 	}
+
+	if err := j.overwriteVolCtxWithPVCAnnotations(ctx, volumeID, volCtx); err != nil {
+		klog.Warningf("Overwrite volCtx with PVC annotations error: %v", err)
```
Comment (timfeirg) [inline pkg/juicefs/juicefs.go]: maybe use a log level of error since this will definitely go against user's expectation?

**Review by timfeirg — APPROVED** [review]

File: pkg/controller/pod_driver.go [hunk 4]
```diff
@@ -376,7 +376,7 @@ func (p *PodDriver) podDeletedHandler(ctx context.Context, pod *corev1.Pod) erro
 				controllerutil.AddFinalizer(newPod, config.Finalizer)
 				klog.Infof("Need to create pod %s %s", pod.Name, pod.Namespace)
 				if err := p.OverwirteMountPodResourcesWithPVC(ctx, newPod); err != nil {
-					klog.Errorf("overwrite mount pod resources wiht pv error %v", err)
+					klog.Errorf("Overwrite mount pod resources wiht pvc error %v", err)
```
Comment (timfeirg) [inline pkg/controller/pod_driver.go]: sorry I mean

```suggestion
					klog.Errorf("Overwrite mount pod resources with pvc error %v", err)
```

**Comment by timfeirg** [comment]
merge and I'll handle the docs update in a separate PR @zxh326

**Pull request merged**